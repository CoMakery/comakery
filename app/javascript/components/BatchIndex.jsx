import React from 'react'
import PropTypes from 'prop-types'
import ProjectSetup from './layouts/ProjectSetup'
import SidebarItem from './styleguide/SidebarItem'
import SidebarItemBold from './styleguide/SidebarItemBold'
import Batch from './Batch'
import Task from './Task'
import styled, { css } from 'styled-components'
import * as Cookies from 'js-cookie'

const Wrapper = styled.div`
`

const Text = styled.div`
  font-family: Georgia;
  font-size: 14px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.29;
  letter-spacing: normal;

  ul {
    list-style-type: decimal;
  }

  a {
    color: #0089f4;
    text-decoration: none;
  }

  a:hover,
  a:focus {
    text-decoration: underline;
  }
`

const Title = styled.div`
  display: flex;
  align-items: center;
  margin-bottom: 7px;
  margin-top: 7px;
`

const TitleText = styled.div`
  margin-bottom: 3px;
  font-size: 15px;
  font-weight: 600;
  text-transform: uppercase;
`

const BatchStyled = styled(Batch)`
  margin-bottom: 20px;
`

const Filter = styled.span`
  margin-bottom: 3px;
  margin-right: 15px;
  font-size: 14px;
  font-weight: 600;
  text-transform: uppercase;
  color: #9b9b9b;

  &:hover {
    cursor: pointer;
    text-decoration: underline;
  }

  ${props => props.filter === 'ready' && props.selected && css`
    color: #4a4a4a;
  `}

  ${props => props.filter === 'started' && props.selected && css`
    color: #008e9b;
  `}

  ${props => props.filter === 'submitted' && props.selected && css`
    color: #007ae7;
  `}

  ${props => props.filter === 'accepted' && props.selected && css`
    color: #5037f7;
  `}

  ${props => props.filter === 'paid' && props.selected && css`
    color: #fb40e5;
  `}

  ${props => props.filter === 'rejected' && props.selected && css`
    color: #ff4d4d;
  `}

  ${props => props.filter === 'cancelled' && props.selected && css`
    color: #ff4d4d;
  `}

  ${props => !props.filter && props.selected && css`
    color: #3a3a3a;
  `}
`

const Tasks = styled.div`
  margin: 20px 0;
`

const CreateTaskButton = styled.div`
  font-family: Montserrat;
  font-size: 14px;
  font-weight: bold;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: normal;
  text-align: right;
  text-transform: uppercase;
  margin: 30px 0px;

  a {
    color: #0089f4;
    text-decoration: none;
  }

  a:hover,
  a:focus {
    text-decoration: underline;
  }
`

class BatchIndex extends React.Component {
  constructor(props) {
    super(props)
    this.handleListClick = this.handleListClick.bind(this)
    this.state = {
      selectedBatch     : null,
      selectedTaskFilter: null
    }
  }

  componentDidMount() {
    this.setState({
      selectedBatch: this.props.batches.find(batch => String(batch.id) === Cookies.get('selectedBatchId'))
    })
  }

  handleListClick(batch) {
    this.setState({
      selectedBatch: batch
    })
    Cookies.set('selectedBatchId', batch.id)
  }

  render() {
    return (
      <React.Fragment>
        <ProjectSetup
          className="batch-index"
          projectId={this.props.project.id}
          projectTitle={this.props.project.title}
          projectPage="batches"
          sidebar={
            <React.Fragment>
              <div className="batch-index--sidebar">
                <SidebarItemBold
                  className="batch-index--sidebar--item__bold"
                  iconLeftName="BATCH/WHITE.svg"
                  iconRightName="PLUS.svg"
                  text="Create a New Batch"
                  onClick={(_) => window.location = this.props.newBatchPath}
                />

                {this.props.batches.length > 0 &&
                  <React.Fragment>
                    <hr className="batch-index--sidebar--hr" />

                    {this.props.batches.map((b, i) =>
                      <SidebarItem
                        className="batch-index--sidebar--item"
                        key={i}
                        iconLeftName="BATCH/ACTIVE.GRADIENT.svg"
                        text={b.name}
                        notificationColor={b.published ? 'green' : 'orange'}
                        selected={this.state.selectedBatch === b}
                        onClick={(_) => this.handleListClick(b)}
                      />
                    )}
                  </React.Fragment>
                }
              </div>
            </React.Fragment>
          }
        >
          {this.props.batches.length === 0 &&
            <Text>
              <ul>
                <li>Batches are “folders” that groups of similar tasks reside in.</li>
                <li>Projects can have one or many batches in them.</li>
                <li>A batch must be <a href={this.props.newBatchPath}>created</a> before a task can be created.</li>
              </ul>
            </Text>
          }
          {this.state.selectedBatch &&
            <React.Fragment>
              <Wrapper>
                <Title>
                  <TitleText>batch</TitleText>
                </Title>

                <BatchStyled batch={this.state.selectedBatch} />

                <CreateTaskButton>
                  <a href={this.state.selectedBatch.newTaskPath}>
                    create a task +
                  </a>
                </CreateTaskButton>

                <Tasks>
                  {this.state.selectedBatch.tasks.length > 0 &&
                    <Title>
                      <Filter filter={this.state.selectedTaskFilter} selected={!this.state.selectedTaskFilter} onClick={(_) => this.setState({selectedTaskFilter: null})}>all tasks</Filter>
                      <Filter filter={this.state.selectedTaskFilter} selected={this.state.selectedTaskFilter === 'ready'} onClick={(_) => this.setState({selectedTaskFilter: 'ready'})}>ready</Filter>
                      <Filter filter={this.state.selectedTaskFilter} selected={this.state.selectedTaskFilter === 'started'} onClick={(_) => this.setState({selectedTaskFilter: 'started'})}>started</Filter>
                      <Filter filter={this.state.selectedTaskFilter} selected={this.state.selectedTaskFilter === 'submitted'} onClick={(_) => this.setState({selectedTaskFilter: 'submitted'})}>submitted</Filter>
                      <Filter filter={this.state.selectedTaskFilter} selected={this.state.selectedTaskFilter === 'accepted'} onClick={(_) => this.setState({selectedTaskFilter: 'accepted'})}>accepted</Filter>
                      <Filter filter={this.state.selectedTaskFilter} selected={this.state.selectedTaskFilter === 'paid'} onClick={(_) => this.setState({selectedTaskFilter: 'paid'})}>paid</Filter>
                      <Filter filter={this.state.selectedTaskFilter} selected={this.state.selectedTaskFilter === 'rejected'} onClick={(_) => this.setState({selectedTaskFilter: 'rejected'})}>rejected</Filter>
                      <Filter filter={this.state.selectedTaskFilter} selected={this.state.selectedTaskFilter === 'cancelled'} onClick={(_) => this.setState({selectedTaskFilter: 'cancelled'})}>cancelled</Filter>
                    </Title>
                  }
                  {this.state.selectedBatch.tasks.filter(task => !this.state.selectedTaskFilter || task.status === this.state.selectedTaskFilter).map((t, i) =>
                    <Task key={i} task={t} />
                  )}
                </Tasks>
              </Wrapper>
            </React.Fragment>
          }
        </ProjectSetup>
      </React.Fragment>
    )
  }
}

BatchIndex.propTypes = {
  batches     : PropTypes.array.isRequired,
  newBatchPath: PropTypes.string,
  project     : PropTypes.object
}
BatchIndex.defaultProps = {
  batches     : [],
  newBatchPath: '',
  project     : {}
}
export default BatchIndex
