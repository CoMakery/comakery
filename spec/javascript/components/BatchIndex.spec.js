import React from 'react'
import { mount } from 'enzyme'
import BatchIndex from 'components/BatchIndex'

describe('BatchIndex', () => {
  it('renders correctly without props', () => {
    const wrapper = mount(<BatchIndex />)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.exists('.batch-index')).toBe(true)
    expect(wrapper.exists('.batch-index--sidebar')).toBe(true)
    expect(wrapper.exists('.batch-index--sidebar SidebarItemBold')).toBe(true)
    expect(wrapper.find('.batch-index--sidebar SidebarItemBold').props().iconLeftName).toBe('BATCH/WHITE.svg')
    expect(wrapper.find('.batch-index--sidebar SidebarItemBold').props().iconRightName).toBe('PLUS.svg')
    expect(wrapper.exists('.batch-index--sidebar--info')).toBe(false)
  })

  it('renders correctly with batches', () => {
    const batches = [
      {
        'id'         : 10,
        'projectId'  : 3,
        'name'       : 'Write a book',
        'description': 'Write a book',
        'goal'       : 'Write a book',
        'specialty'  : 'writing',
        'editPath'   : '/projects/3/batches/10/edit',
        'destroyPath': '/projects/3/batches/10',
        'newTaskPath': '/projects/3/batches/10/tasks/new',
        'tasks'      : [
          {
            'id'          : 21,
            'description' : 'Chapter VII',
            'name'        : 'Chapter VII',
            'why'         : 'Chapter VII',
            'requirements': 'Chapter VII',
            'status'      : 'ready',
            'amount'      : '1000.0',
            'tokenSymbol' : 'DMT',
            'awardPath'   : '/projects/3/batches/10/tasks/21',
            'editPath'    : '/projects/3/batches/10/tasks/21/edit',
            'destroyPath' : '/projects/3/batches/10/tasks/21',
            'contributor' : {
              'name' : 'test',
              'image': 'http://test'
            }
          },
          {
            'id'          : 22,
            'description' : 'Chapter VIII',
            'quantity'    : '1.21',
            'totalAmount' : '1210.0',
            'name'        : 'Chapter VIII',
            'why'         : 'Chapter VIII',
            'requirements': 'Chapter VIII',
            'status'      : 'done',
            'amount'      : '100.0',
            'tokenSymbol' : 'DMT',
            'awardPath'   : '/projects/3/batches/10/tasks/21',
            'editPath'    : '/projects/3/batches/10/tasks/21/edit',
            'destroyPath' : '/projects/3/batches/10/tasks/21',
            'contributor' : {
              'name' : 'test',
              'image': 'http://test'
            }
          }
        ]
      }
    ]

    const wrapper = mount(<BatchIndex batches={batches} />)

    expect(wrapper.exists('SidebarItem[text="Write a book"]')).toBe(true)
  })

  it('displays correct token details on sidebar item click', () => {
    const batches = [
      {
        'id'         : 10,
        'projectId'  : 3,
        'name'       : 'Write a book',
        'description': 'Write a book',
        'goal'       : 'Write a book',
        'specialty'  : 'writing',
        'editPath'   : '/projects/3/batches/10/edit',
        'destroyPath': '/projects/3/batches/10',
        'newTaskPath': '/projects/3/batches/10/tasks/new',
        'tasks'      : [
          {
            'id'          : 21,
            'description' : 'Chapter VII',
            'name'        : 'Chapter VII',
            'why'         : 'Chapter VII',
            'requirements': 'Chapter VII',
            'status'      : 'ready',
            'amount'      : '1000.0',
            'tokenSymbol' : 'DMT',
            'awardPath'   : '/projects/3/batches/10/tasks/21',
            'editPath'    : '/projects/3/batches/10/tasks/21/edit',
            'destroyPath' : '/projects/3/batches/10/tasks/21',
            'contributor' : {
              'name' : 'test',
              'image': 'http://test'
            }
          },
          {
            'id'          : 22,
            'description' : 'Chapter VIII',
            'quantity'    : '1.21',
            'totalAmount' : '1210.0',
            'name'        : 'Chapter VIII',
            'why'         : 'Chapter VIII',
            'requirements': 'Chapter VIII',
            'status'      : 'done',
            'amount'      : '100.0',
            'tokenSymbol' : 'DMT',
            'awardPath'   : '/projects/3/batches/10/tasks/21',
            'editPath'    : '/projects/3/batches/10/tasks/21/edit',
            'destroyPath' : '/projects/3/batches/10/tasks/21',
            'contributor' : {
              'name' : 'test',
              'image': 'http://test'
            }
          }
        ]
      },
      {
        'id'         : 11,
        'projectId'  : 3,
        'name'       : 'Write a blog post',
        'description': 'Write a blog post',
        'goal'       : 'Write a blog post',
        'specialty'  : 'writing',
        'editPath'   : '/projects/3/batches/11/edit',
        'destroyPath': '/projects/3/batches/11',
        'newTaskPath': '/projects/3/batches/11/tasks/new',
        'tasks'      : []
      }
    ]

    const wrapper = mount(<BatchIndex batches={batches} />)

    wrapper.find('SidebarItem[text="Write a book"]').simulate('click')
    expect(wrapper.find('.layout--content--content__sidebared').text()).toMatch(/Write a book/)

    wrapper.find('SidebarItem[text="Write a blog post"]').simulate('click')
    expect(wrapper.find('.layout--content--content__sidebared').text()).toMatch(/Write a blog post/)
    expect(wrapper.find('.layout--content--content__sidebared').text()).not.toMatch(/Write a book/)
  })
})
