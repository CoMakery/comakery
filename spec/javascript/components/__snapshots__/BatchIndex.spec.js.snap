// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchIndex renders correctly without props 1`] = `
<BatchIndex
  batches={Array []}
  newBatchPath=""
  project={Object {}}
>
  <ProjectSetup
    className="batch-index"
    projectId={null}
    projectPage="batches"
    projectTitle=""
    sidebar={
      <React.Fragment>
        <div
          className="batch-index--sidebar"
        >
          <SidebarItemBold
            className="batch-index--sidebar--item__bold"
            iconLeftName="BATCH/WHITE.svg"
            iconRightName="PLUS.svg"
            onClick={[Function]}
            text="Create a New Batch"
          />
        </div>
      </React.Fragment>
    }
    subfooter={null}
  >
    <Layout
      category=""
      className="project-setup batch-index"
      customTitle={
        <ProjectSetupHeader
          projectId={null}
          projectOwner={true}
          projectPage="batches"
          projectTitle=""
        />
      }
      hasBackButton={false}
      sidebar={
        <React.Fragment>
          <div
            className="batch-index--sidebar"
          >
            <SidebarItemBold
              className="batch-index--sidebar--item__bold"
              iconLeftName="BATCH/WHITE.svg"
              iconRightName="PLUS.svg"
              onClick={[Function]}
              text="Create a New Batch"
            />
          </div>
        </React.Fragment>
      }
      subfooter={null}
      title={null}
    >
      <div
        className="layout project-setup batch-index"
      >
        <div
          className="layout--content"
        >
          <div
            className="layout--content--title"
          >
            <ProjectSetupHeader
              projectId={null}
              projectOwner={true}
              projectPage="batches"
              projectTitle=""
            >
              <div
                className="layout--content--title--header"
              />
              <span
                className="layout--content--title--nav"
              >
                <a
                  href="#"
                >
                  project settings
                </a>
              </span>
              <span
                className="layout--content--title--nav__current"
              >
                <a
                  href="#"
                >
                  batches & tasks
                </a>
              </span>
              <span
                className="layout--content--title--nav"
              >
                <a
                  href="#"
                >
                  contributors
                </a>
              </span>
              <span
                className="layout--content--title--nav"
              >
                <a
                  href="#"
                >
                  payments
                </a>
              </span>
              <span
                className="layout--content--title--nav"
              >
                <a
                  href="#"
                >
                  overview
                </a>
              </span>
            </ProjectSetupHeader>
          </div>
          <hr
            className="layout--content--hr"
          />
          <div
            className="layout--content--wrapper"
          >
            <div
              className="layout--content--sidebar"
            >
              <div
                className="batch-index--sidebar"
              >
                <SidebarItemBold
                  className="batch-index--sidebar--item__bold"
                  iconLeftName="BATCH/WHITE.svg"
                  iconRightName="PLUS.svg"
                  onClick={[Function]}
                  text="Create a New Batch"
                >
                  <SidebarItem
                    className="sidebar-item__bold batch-index--sidebar--item__bold"
                    iconLeftName="BATCH/WHITE.svg"
                    iconLeftUrl=""
                    iconRightName="PLUS.svg"
                    onClick={[Function]}
                    selected={false}
                    text="Create a New Batch"
                  >
                    <div
                      className="sidebar-item--wrapper"
                      onClick={[Function]}
                    >
                      <div
                        className="sidebar-item sidebar-item__bold batch-index--sidebar--item__bold"
                      >
                        <span
                          className="sidebar-item--icon-left"
                        >
                          <Icon
                            className=""
                            name="BATCH/WHITE.svg"
                          >
                            <img
                              className="icon icon__BATCH/WHITE-svg"
                              src="test-file-stub"
                            />
                          </Icon>
                        </span>
                        <span
                          className="sidebar-item--text"
                        >
                          Create a New Batch
                        </span>
                        <span
                          className="sidebar-item--icon-right"
                        >
                          <Icon
                            className=""
                            name="PLUS.svg"
                          >
                            <img
                              className="icon icon__PLUS-svg"
                              src="test-file-stub"
                            />
                          </Icon>
                        </span>
                        <div
                          className="sidebar-item--left-border"
                        />
                      </div>
                    </div>
                  </SidebarItem>
                </SidebarItemBold>
              </div>
            </div>
            <div
              className="layout--content--content__sidebared"
            >
              <styled.div>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-cMljjf",
                        "isStatic": true,
                        "lastClassName": "kPwhoC",
                        "rules": Array [
                          "
  font-family: Georgia;
  font-size: 14px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.29;
  letter-spacing: normal;

  ul {
    list-style-type: decimal;
  }

  a {
    color: #0089f4;
    text-decoration: none;
  }

  a:hover,
  a:focus {
    text-decoration: underline;
  }
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-cMljjf",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="sc-cMljjf kPwhoC"
                  >
                    <ul>
                      <li>
                        Batches are “folders” that groups of similar tasks reside in.
                      </li>
                      <li>
                        Projects can have one or many batches in them.
                      </li>
                      <li>
                        A batch must be 
                        <a
                          href=""
                        >
                          created
                        </a>
                         before a task can be created.
                      </li>
                    </ul>
                  </div>
                </StyledComponent>
              </styled.div>
            </div>
          </div>
        </div>
      </div>
    </Layout>
  </ProjectSetup>
</BatchIndex>
`;
